### PR TITLE
feat: implement very crude and bare-bones RSS feed

### LINKED ISSUE #1673: An Atom (RSS) feed for notifications (including the public)
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

New Notification

### 🔖 Feature description

Add [RSS](https://en.wikipedia.org/wiki/RSS)/[Atom](https://en.wikipedia.org/wiki/Atom_(Web_standard)) feeds as a way of notifications.

### ✔️ Solution

Add a new notification method ([RSS](https://en.wikipedia.org/wiki/RSS)/[Atom](https://en.wikipedia.org/wiki/Atom_(Web_standard)) feeds).

Also, add an option of which feed should correspond to a status page on when modifying the status page.

### ❓ Alternatives

_No response_

### 📝 Additional Context

Some examples of real world use:
- User expressing how they used the RSS feeds (or Atom feeds) available: https://github.com/louislam/uptime-kuma/pull/124#issuecomment-924280278
- [Instatus](https://instatus.com/) example: https://status.sponsor.ajay.app/ (see "Get updates" button) ([Atom feed for SponsorBlock](https://sponsorblock.instatus.com/history.atom))
- [My website](https://jacksonchen666.com/feeds/)
- A client that can do RSS stuff: [Thunderbird](https://www.thunderbird.net/) ([help page](https://support.mozilla.org/en-US/kb/how-subscribe-news-feeds-and-blogs))

### PR DESCRIPTION
**Pull request by MrYakobo**
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description

Fixes #1673.

This PR adds RSS capabilities to status-pages without extra configuration.
It's implemented as a custom route: append `/rss` to any status URL to get the RSS feed.
I've checked out the discussions in #2129 and #1673.
This PR adds a new server-side dependency `feed`.

Right now, the RSS feed looks like this:

```bash
curl http://localhost:3001/status/abc/rss

<?xml version="1.0" encoding="utf-8"?>
<rss version="2.0">
    <channel>
        <title>uptime kuma rss feed</title>
        <link>http://localhost:3001/status/abc</link>
        <description>feed for monitors that are down</description>
        <lastBuildDate>Sat, 24 Aug 2024 21:55:21 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>https://github.com/jpmonette/feed</generator>
        <language>en</language>
        <item>
            <title><![CDATA[myserver is down]]></title>
            <pubDate>Sat, 24 Aug 2024 19:55:05 GMT</pubDate>
            <description><![CDATA[myserver has been down since 2024-08-24 21:55:05.738]]></description>
        </item>
    </channel>
</rss>
```

Right now, the RSS feed is missing this information that exists on the regular status page:

- The label "All Systems Operational"
- Description of status page
- Percentage of uptime for each service

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [x] My changes generates no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

Please do not use any external image service. Instead, just paste in or drag and drop the image here, and it will be uploaded automatically.

### CODE REVIEW

File: server/model/status_page.js
Comment (CommanderStorm) [inline server/model/status_page.js:172]: nit: can be simplified

```suggestion
            for (const monitor of monitorGroup.monitorList) {
                const heartbeat = await R.findOne("heartbeat", "monitor_id = ? ORDER BY time DESC", [ monitor.id ]);
                if (heartbeat && !heartbeat.status) {
                    heartbeats.push({
                        ...monitor,
                        status: heartbeat.status,
                        time: heartbeat.time,
                    });
```

File: server/model/status_page.js
Comment (CommanderStorm) [inline server/model/status_page.js:170]: FYI: This is fine here, because we cache the response for 5m, but usually this would be a N+1 query.
See https://github.com/louislam/uptime-kuma/pull/5025 for a realted PR.

I have not measured the performance impact of this PR on said endpoint as I currently am quite swamped with other work. If you can refactor this, that would be nice.
If you can't, so be it ^^

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
Good job 👍🏻 

I have left a few inline comments

File: server/model/status_page.js
Comment (MrYakobo) [reply server/model/status_page.js:29]: I've written this method like `handleStatusPageResponse`, which doesn't have a try-catch. However, I forgot to handle the case of statusPage being null. Will fix. :+1:

Comment (MrYakobo) [reply server/model/status_page.js:29]: The function handleStatusPageResponse doesn't do that. However, I forgot to handle the case when slug is not found. Will fix :+1:

File: server/model/status_page.js
Comment (MrYakobo) [reply server/model/status_page.js:80]: I found this code in StatusPage.vue that interprets the status field from the database:

```js
            let status = STATUS_PAGE_ALL_UP;
            let hasUp = false;

            for (let id in this.$root.publicLastHeartbeatList) {
                let beat = this.$root.publicLastHeartbeatList[id];

                if (beat.status === MAINTENANCE) {
                    return STATUS_PAGE_MAINTENANCE;
                } else if (beat.status === UP) {
                    hasUp = true;
                } else {
                    status = STATUS_PAGE_PARTIAL_DOWN;
                }
            }

            if (! hasUp) {
                status = STATUS_PAGE_ALL_DOWN;
            }

            return status;
```

I'll do something similar for the RSS feed :+1:

Comment (MrYakobo) [reply server/model/status_page.js:80]: To your questions,

> That is just the last heartbeat that failed.
> That description however can be misinterpreted as the first DOWN heartbeat (=> the total downtime for said heatbeat).

in getRSSPageData, I'm filtering for the last heartbeat of monitor_id=?. If that heartbeat is falsy, its added to the RSS feed. When the monitor is UP again, the RSS feed won't publish it anymore. I think this is the desired behaviour.

File: server/model/status_page.js
Comment (MrYakobo) [reply server/model/status_page.js:170]: You're fully right. To avoid the N+1, I guess I'd have to surgically pick out the code for groupBean.toPublicJSON and add the relevant info in a JOIN. Not too confident in this code base to do that 😄 

And as you say, the caching mitigates this issue a little.

File: server/model/status_page.js
Comment (MrYakobo) [reply server/model/status_page.js]: Fixed in chore commit

File: server/model/status_page.js
Comment (CommanderStorm) [reply server/model/status_page.js:87]: that sounds fine (I don't use RSS-Feeds, your call).

Have you tested that a falsy heartbeat means what you think it does?
Only the following values are falsy: https://developer.mozilla.org/en-US/docs/Glossary/Falsy

=> you are not filtering for the things you think you are..

Comment (MrYakobo) [reply server/model/status_page.js:87]: Great call mate. I've not worked with js for many years, so I'm a little bit rusty on these quirks :+1:

**Comment by MrYakobo** [comment]
Hello again! I think I've addressed your comments @CommanderStorm. What's hindering us from getting this PR merged? :+1:

**Comment by CommanderStorm** [comment]
Blocking merging is Amongst others that my last review https://github.com/louislam/uptime-kuma/pull/5047#discussion_r1731567726 is not addressed as far as I can tell..

You are still filtering for truthy-ness and thus likely displaying `UP` heartbeats as down.

**Comment by MrYakobo** [comment]
@CommanderStorm that's in my opinion addressed in commit [35349a](https://github.com/louislam/uptime-kuma/pull/5047/commits/35349aa97c84b8f2c29e0fee0ef031a81261e6d7) , where I explicitly look for DOWN beats.

What do you mean with "Amongst others"? Are there more unanswered comments?

**Review by CommanderStorm — APPROVED** [review]
Sorry, I could not do a proper review yesterday because of a terrible internet connection in Shanghai (firewall and stuffs you know).
(I flew onto vacation to Tokio)

LGTM now, I overlooked the reason why this is not filtered below.

Merging with junior-approval

**Comment by MrYakobo** [comment]
Thank you @CommanderStorm for your guidance in this PR! Happy vacation!

**Pull request merged**
